lib/src/transformer/utils.dart

```diff
@@ -0,0 +1,224 @@
+
+class Replacement {
+  final SourceSpan span;
+  final String newText;
+
+  Replacement(this.span, this.newText);
+
+  @override
+  String toString() => 'Replacement (span: $span, newText: $newText)';
+}
+
+class TransformedSourceFile {
+  final List<Replacement> _replacements = [];
+
+  final SourceFile sourceFile;
+
+  TransformedSourceFile(this.sourceFile);
+
+  bool get isModified => _replacements.isNotEmpty;
+
+  void insert(SourceLocation location, String text) {
+    _replacements.add(
+        new Replacement(location.pointSpan(), text)
+    );
+  }
+
+  void replace(SourceSpan span, String text) {
+    _replacements.add(
+        new Replacement(span, text)
+    );
+  }
+
+  void iterateReplacements({onUnmodified(String string), onRemoval(String string), onAddition(String string)}) {
+    _replacements.sort((r1, r2) => r1.span.compareTo(r2.span));
+
+    var lastEdge = 0;
+    for (Replacement replacement in _replacements) {
+      if (replacement.span.start.offset < lastEdge) {
+        throw 'Overlapping replacement $replacement in replacements $_replacements.';
+      }
+
+      var unmodifiedText = sourceFile.getText(lastEdge, replacement.span.start.offset);
+      var removalText = replacement.span.text;
+      var additionText = replacement.newText;
+
+      if (onUnmodified != null && unmodifiedText.isNotEmpty) {
+        onUnmodified(unmodifiedText);
+      }
+      if (onRemoval != null && removalText.isNotEmpty) {
+        onRemoval(removalText);
+      }
+      if (onAddition != null && additionText.isNotEmpty) {
+        onAddition(additionText);
+      }
+
+      lastEdge = replacement.span.end.offset;
+    }
+
+    var unmodifiedText = sourceFile.getText(lastEdge);
+    if (onUnmodified != null && unmodifiedText != null) {
+      onUnmodified(unmodifiedText);
+    }
+  }
+
+  String getTransformedText() {
+    StringBuffer transformedSource = new StringBuffer();
+
+    iterateReplacements(
+        onUnmodified: transformedSource.write,
+        onAddition: transformedSource.write
+    );
+
+    return transformedSource.toString();
+  }
+
+  String getHtmlDiff() {
+    const HtmlEscape elementEscaper = const HtmlEscape(HtmlEscapeMode.ELEMENT);
+    const HtmlEscape attrEscaper = const HtmlEscape(HtmlEscapeMode.ATTRIBUTE);
+
+    StringBuffer diff = new StringBuffer();
+
+    void writeDiff(String source, String className) {
+      diff.write('<span class="$className">');
+      diff.write(elementEscaper.convert(source));
+      diff.write('</span>');
+    }
+
+    iterateReplacements(
+        onUnmodified: (source) => writeDiff(source, 'diff-unmodified'),
+        onRemoval: (source) => writeDiff(source, 'diff-removal'),
+        onAddition: (source) => writeDiff(source, 'diff-addition')
+    );
+
+    return '''
+        <!DOCTYPE html>
+        <html>
+        <head>
+          <title>web_skin_dart Transformer Diff - ${attrEscaper.convert(sourceFile.url.path)}</title>
```

Comment:
Splitting it out should be really straightforward. I've got some small additions to `instantiateAnnotation` landing in a ticket I'm about to PR, but otherwise, it should be pretty much finalized.

I just need help thinking of a name, and then I can make a JIRA ticket for the repo.

`w_codegen_util`? `transformer_codegen_util`?

Thoughts?